### DISCUSSION TITLE
Simple third person controller of a space ship using OimoPhysics, Three.js, and JavaScript

### DISCUSSION THREAD

**Discussion by ivan-enzhaev** [saharan/OimoPhysics#74] [Show and tell]
Playground Plunker: https://plnkr.co/edit/ENV8AfwIztG5YS7g?preview

Source code on GitHub: https://github.com/8Observer8/ship-movement-oimophysics-threejs-js

Updated 3/25/2026 - It was created without AI.

![ship-movement-oimophysics-rollup-threejs-js-original](https://github.com/saharan/OimoPhysics/assets/3908473/c15e7851-4eda-4927-8355-1e707cf7bc39)

**Comment by ivan-enzhaev** [comment]
Gemini 3 said that I should protect `delta time` and I made it. I haved applied the `1. Clamping the Delta Time` case in the Plunker Playground and GitHub. https://github.com/8Observer8/ship-movement-oimophysics-threejs-js/commit/88202d324a0bf022e151483cb25a613783b238d8

Gemini 3 answer:

In game development, "protecting" `dt` (delta time) refers to preventing physics glitches that occur when the frame rate spikes or drops significantly. If the user tabs away or the browser stutters, `dt` can become a large number, causing your physics objects to tunnel through walls or fly off the map.

Here is how you can protect it in your Three.js and OimoPhysics setup:

### 1. Clamping the Delta Time

The simplest protection is to cap the maximum value `dt` can reach. If a frame takes longer than, say, **1/10th of a second**, you force the physics engine to treat it as a smaller slice of time.

```javascript
// Inside your render() function
dt = clock.getDelta();

// Protect dt: Cap it at 0.1 seconds (10 FPS equivalent)
const maxDt = 0.1;
if (dt > maxDt) dt = maxDt;

world.step(dt);

```

---

### 2. Implementing a Fixed Time Step

Physics engines like Oimo perform best with a **fixed** time step. Instead of passing a variable `dt` directly into `world.step()`, you accumulate time and run the physics simulation in consistent increments (e.g., 60 times per second).

Replace your current physics update logic with this "accumulator" pattern:

```javascript
let accumulator = 0;
const fixedTimeStep = 1 / 60; // 60Hz physics

function render() {
    keyboardHandler();

    const frameTime = clock.getDelta();
    // Cap frameTime to prevent "Spiral of Death" if the game lags heavily
    accumulator += Math.min(frameTime, 0.25); 

    // Step the physics world in fixed increments
    while (accumulator >= fixedTimeStep) {
        world.step(fixedTimeStep);
        accumulator -= fixedTimeStep;
    }

    // ... rest of your rendering code
}

```

---

### 3. Why this matters for your code

Looking at your current implementation, there are two specific risks:

* **Tunnelling:** In `applyThirdPersonInput`, you are setting linear velocity. If `dt` is very large, Oimo will calculate the new position as . If  is high enough, your ship will teleport to the other side of your "wall" in a single frame.
* **Browser Backgrounding:** When a user switches browser tabs, `requestAnimationFrame` pauses, but `THREE.Clock` keeps running. When they switch back, the first `dt` will be the total time they were away (potentially minutes), which will crash the physics simulation without a cap.

### 4. Small Bug Fix in `applyThirdPersonInput`

I noticed you are clearing the `input` array inside the function, but it is declared as a `let` globally. To ensure it clears correctly and doesn't accumulate indefinitely if the function isn't called, make sure you reset it to an empty array:

```javascript
// Change this at the end of applyThirdPersonInput
if (inputs.length) {
    input.length = 0; // Better way to clear the array reference
}

```

**Would you like me to show you how to interpolate the mesh position so the movement looks smooth even when the physics step is fixed?**

**Reply by JorkDev** [reply]
A few things worth addressing:

**Clamping `dt`** treats the symptom, a variable time step still enters a solver that expects consistency, jitter and tunnelling don't disappear, they just become harder to reproduce.

**The accumulator pattern** is incomplete. Without lerping by `accumulator / fixedTimeStep` before the render pass, stutter is inevitable when the display refresh rate doesn't align with the fixed step, 144Hz + 60Hz physics being the obvious case.

**Browser backgrounding** is a non-issue. `THREE.Clock.getDelta()` resets on each call, and modern browsers already throttle the first `requestAnimationFrame` callback after a tab switch.

Lastly, if Gemini authored this, you're not solving your problem, you're inheriting someone else's assumptions. A custom controller only has value if the reasoning behind it is yours.

**Reply by ivan-enzhaev** [reply]
Thank you very much! No, It is my code. I just copied my code to Gemini to check it. I will research.
